### DISCUSSION TITLE
Custom Combo box with a button.

### DISCUSSION THREAD

**Discussion by BrunoDenisov** [AvaloniaUI/Avalonia#16933] [Q&A] [answered]
Hello, I'm trying to create a combo box that at the bottom(or top) of its dropdown menu has a button to where a user would be able to add more items with the button. I've tired a few different ways, Making it as a resource, making custom user control but I can't seem to get it to work. If anyone could point me to some resources that do a similiar  thing I would apriciate it.

**Comment by BrunoDenisov** [comment]
```
<ComboBox Margin="6"
      ItemsSource="{Binding FileNames}"
      SelectedItem="{Binding SelectedFileName}">
<ComboBox.ItemsPanel>
    <ItemsPanelTemplate>
        <StackPanel>
            <ItemsPresenter />
            <Button Content="Add" Command="{Binding AddItemCommand}" />
        </StackPanel>
    </ItemsPanelTemplate>
</ComboBox.ItemsPanel>
</ComboBox>
```

this is my latest attempt doing it directly in another user contorol

**Reply by BrunoDenisov** [reply]
I've worked out that an Itempresenter takes up the hole stack panel and I;m not sure how to solve it

**Reply by timunie** [reply, marked as answer]
You need to change entire Control template. In source code you can copy and modify the original one.

**Reply by BrunoDenisov** [reply]
Right I'll try that, for now I just dida hack solution, with a combo box and one otpiton that is always last. But will see if I can improve it.

**Reply by Grave18** [reply]
> You need to change entire Control template. In source code you can copy and modify the original one.

What class did you mean?

**Reply by timunie** [reply]
If you are using Fluent Theme, this one: https://github.com/AvaloniaUI/Avalonia/blob/c66008b627926b0cd35abcc83d1bf681e416b5a8/src/Avalonia.Themes.Fluent/Controls/ComboBox.xaml#L57-L141

**Reply by Grave18** [reply]
```xaml
<ItemsPresenter Name="PART_ItemsPresenter" 
                Margin="{DynamicResource ComboBoxDropdownContentMargin}" 
                ItemsPanel="{TemplateBinding ItemsPanel}" /> 
```

I dug into the code. Tell me, am I in the right direction? As I understand it, ItemsPresenter is responsible for displaying in Popup.
Maybe I must inherit ItemPresenter base class or something?

**Reply by timunie** [reply]
It is 

```txt
Popup
  - ScrollViewer
     - ItemsPresenter
``` 

you can do:
```
```txt
Popup
   - Grid
     - ScrollViewer
       - ItemsPresenter
     - Button
``` 

```txt
Popup
   - ScrollViewer
     - StackPanel
       - ItemsPresenter
       - Button
```

**Reply by Grave18** [reply]
Thanks!

**Comment by GamerGirlandCo** [comment]
```cs
public class MyComboBox : ComboBox {
   protected override System.Type StyleKeyOverride => typeof(ComboBox)
  // rest of your code...
}
```
